### DISCUSSION TITLE
Connecting to Azure Database for MySQL server on serverless Vercel API

### DISCUSSION THREAD

**Discussion by yannickgloster** [prisma/prisma#11453] [Q&A] [answered]
Following #1673, I converted the SSL Certificate that I needed from here [Azure Database for MySQL](https://docs.microsoft.com/en-us/azure/mysql/howto-configure-ssl) into base64 to store it in as an environment variable that I can then load in into the tmp folder on the serverless function. 
I do this when I create my prisma client:
`/lib/prisma.js`
```js
import { PrismaClient } from "@prisma/client";

const fs = require("fs");
const path = "/tmp";

try {
  if (fs.existsSync(`${path}/BaltimoreCyberTrustRoot.crt.pem`)) {
    // do nothing
  } else {
    // creates crt.pem file from ENV

    fs.mkdirSync(path, { recursive: true });

    fs.writeFileSync(
      `${path}/BaltimoreCyberTrustRoot.crt.pem`,
      Buffer.from(process.env.CRT_PEM, "base64").toString("ascii")
    );
  }
} catch (err) {
  console.error(err);
}

const prisma = global.prisma || new PrismaClient();

if (process.env.NODE_ENV === "development") global.prisma = prisma;

export default prisma;
```

Locally this works as expected, and when I check on the API endpoint, the file exists correctly
```js
import fs from "fs";

export default async function handler(req, res) {
  const files = fs.readFileSync(`/tmp/BaltimoreCyberTrustRoot.crt.pem`);

  res.status(200).json({ files: files.toString() });
}
```

My environment variables look like this:
```env
SSL_CERT_PATH=/tmp/BaltimoreCyberTrustRoot.crt.pem
DATABASE_URL=mysql://user:pass@lserverip:3306/registration?sslcert=${SSL_CERT_PATH}&connect_timeout=300
CRT_PEM=longbase64string
```

I'm able to successfully connect to the DB when I use those variables locally, but when deployed I get the following error when I go to make a request to the DB.
```
2022-01-27T23:22:08.847Z	c63d91bb-ae49-40c3-95dd-c20a224d3f4d	ERROR	PrismaClientInitializationError: Can't reach database server at `IP OF SERVER`:`3306`
Please make sure your database server is running at `IP OF SERVER`:`3306`.
    at cb (/var/task/node_modules/@prisma/client/runtime/index.js:38697:17)
    at async handler (/var/task/.next/server/pages/api/tournaments.js:60:25)
    at async Object.apiResolver (/var/task/node_modules/next/dist/server/api-utils.js:102:9)
    at async NextNodeServer.handleApiRequest (/var/task/node_modules/next/dist/server/base-server.js:1076:9)
    at async Object.fn (/var/task/node_modules/next/dist/server/base-server.js:963:37)
    at async Router.execute (/var/task/node_modules/next/dist/server/router.js:222:32)
    at async NextNodeServer.run (/var/task/node_modules/next/dist/server/base-server.js:1103:29)
    at async NextNodeServer.handleRequest (/var/task/node_modules/next/dist/server/base-server.js:319:20)
    at async Server.<anonymous> (/var/task/___next_launcher.js:32:9) {
  clientVersion: '3.8.1',
  errorCode: undefined
}
2022-01-27T23:22:08.847Z	c63d91bb-ae49-40c3-95dd-c20a224d3f4d	ERROR	PrismaClientInitializationError: Can't reach database server at `IP OF SERVER`:`3306`
Please make sure your database server is running at `IP OF SERVER`:`3306`.
    at cb (/var/task/node_modules/@prisma/client/runtime/index.js:38697:17)
    at async handler (/var/task/.next/server/pages/api/tournaments.js:60:25)
    at async Object.apiResolver (/var/task/node_modules/next/dist/server/api-utils.js:102:9)
    at async NextNodeServer.handleApiRequest (/var/task/node_modules/next/dist/server/base-server.js:1076:9)
    at async Object.fn (/var/task/node_modules/next/dist/server/base-server.js:963:37)
    at async Router.execute (/var/task/node_modules/next/dist/server/router.js:222:32)
    at async NextNodeServer.run (/var/task/node_modules/next/dist/server/base-server.js:1103:29)
    at async NextNodeServer.handleRequest (/var/task/node_modules/next/dist/server/base-server.js:319:20)
    at async Server.<anonymous> (/var/task/___next_launcher.js:32:9) {
  clientVersion: '3.8.1',
  errorCode: undefined
}
RequestId: c63d91bb-ae49-40c3-95dd-c20a224d3f4d Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

If I remove the SSL information from my DB connection string, I get an error stating that I need to provide SSL information to continue.
```Server error: `ERROR 28000 (9002): SSL connection is required. Please specify SSL options and retry.\u0000'```

Is there something I'm doing wrong here? 
Thanks in advance

**Comment by stambunan** [comment]
Were you able to find a solution @yannickgloster ? I encountered a similar issue and found using the absolute path of the certificate helped. `/home/site/wwwroot/prisma/cert.pem`.

Perhaps the way Prisma is resolving the relative address is not working in App Services?

**Reply by janpio** [reply]
I think the path `/tmp/BaltimoreCyberTrustRoot.crt.pem` that @yannickgloster was using was the absolute one.

**Reply by yannickgloster** [reply]
Sorry that this wasn't mentioned here. Some environments such as vercel, don't support env composition. Only direct usage of .env files allow for this.

**Comment by yannickgloster** [comment, marked as answer]
Solution Discussed [here](https://github.com/prisma/docs/issues/2835).

Tldr; `.env` composition is only supported directly in `.env` files and not environment variables sent in a system such as vercel.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
